internal/xds/xdsdepmgr/xds_dependency_manager.go

```diff
@@ -130,15 +170,203 @@ func (m *DependencyManager) annotateErrorWithNodeID(err error) error {
 	return fmt.Errorf("[xDS node id: %v]: %v", m.nodeID, err)
 }
 
-// maybeSendUpdateLocked checks that all the resources have been received and sends
-// the current aggregated xDS configuration to the watcher if all the updates
-// are available.
+// maybeSendUpdateLocked checks that all the resources have been received and
+// sends the current aggregated xDS configuration to the watcher if all the
+// updates are available.
 func (m *DependencyManager) maybeSendUpdateLocked() {
-	m.watcher.Update(&xdsresource.XDSConfig{
+	if m.currentListenerUpdate == nil || m.currentRouteConfig == nil {
+		return
+	}
+	config := &xdsresource.XDSConfig{
 		Listener:    m.currentListenerUpdate,
 		RouteConfig: m.currentRouteConfig,
 		VirtualHost: m.currentVirtualHost,
-	})
+		Clusters:    make(map[string]*xdsresource.ClusterResult),
+	}
+	if !EnableClusterAndEndpointsWatch {
+		m.watcher.Update(config)
+		return
+	}
+	edsResourcesSeen := make(map[string]bool)
+	dnsResourcesSeen := make(map[string]bool)
+	clusterResourceSeen := make(map[string]bool)
+	haveAllResources := true
+	for cluster := range m.clustersFromRouteConfig {
+		gotAllResource, _, err := m.populateClusterConfigLocked(cluster, 0, config.Clusters, edsResourcesSeen, dnsResourcesSeen, clusterResourceSeen)
+		if !gotAllResource {
+			haveAllResources = false
+		}
+		if err != nil {
+			config.Clusters[cluster] = &xdsresource.ClusterResult{
+				Err: err,
+			}
+		}
+	}
+
+	// Cancel resources not seen in the tree.
+	for name, endpoint := range m.endpointWatchers {
```

Comment:
Nit: s/endpoint/ep

See: go/go-style/decisions#variable-names
The first line in the above style guideline should give you enough information.